core/src/main/java/org/infinispan/LockedStream.java

```diff
@@ -0,0 +1,104 @@
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Spliterator;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiConsumer;
+import java.util.function.Predicate;
+
+import org.infinispan.configuration.cache.TransactionConfiguration;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.util.function.SerializableBiConsumer;
+import org.infinispan.util.function.SerializablePredicate;
+
+/**
+ * Stream that allows for operation upon data solely with side effects by using {@link LockedStream#forEach(BiConsumer)}
+ * where the <b>BiConsumer</b> is invoked while guaranteeing that the entry being passed is properly locked for the
+ * entire duration of the invocation.
+ * @author wburns
+ * @since 9.1
+ */
+public interface LockedStream<K, V> extends BaseCacheStream<CacheEntry<K, V>, LockedStream<K, V>> {
+   /**
+    * Returns a locked stream consisting of the elements of this stream that match
+    * the given predicate.
+    * <p>
+    * This filter is after the lock is acquired for the given key. This way the filter will see the same value as
+    * the consumer is given.
+    * @param predicate predicate
+    * @return a LockedStream with the filter applied
+    */
+   LockedStream<K, V> filter(Predicate<? super CacheEntry<K, V>> predicate);
+
+   /**
+    * Same as {@link LockedStream#filter(Predicate)} except that the Predicate must also
+    * implement <code>Serializable</code>
+    * <p>
+    * The compiler will pick this overload for lambda parameters, making them <code>Serializable</code>
+    * @param predicate the predicate to filter out unwanted entries
+    * @return a LockedStream with the filter applied
+    */
+   default LockedStream<K, V> filter(SerializablePredicate<? super CacheEntry<K ,V>> predicate) {
+      return filter((Predicate<? super CacheEntry<K, V>> ) predicate);
+   }
+
+   /**
+    * Performs an action for each element of this stream on the primary owner of the given key.
+    * <p>
+    * This method is performed while holding exclusive lock over the given entry and will be released
+    * only after the consumer has completed. If the entry is directly modified via the
+    * {@link java.util.Map.Entry#setValue(Object)} method this will be the same as if
+    * {@link java.util.Map#put(Object, Object)} was invoked.
+    * <p>
+    * If using a pessimistic transaction this lock is not held using a transaction and thus the user can start a
+    * transaction in this consumer which also must be completed before returning. A transaction can be started in
+    * the consumer and if done it will share the same lock used to obtain the key.
+    * <p>
+    * Some methods on the provided cache may not work as expected. These include
+    * {@link AdvancedCache#putForExternalRead(Object, Object)}, {@link AdvancedCache#lock(Object[])},
+    * {@link AdvancedCache#lock(Collection)}, {@link AdvancedCache#removeGroup(String)}, all of the async methods on
```

Comment:
Then it's not different from the incomplete user transaction - user has to make sure it waits for all asynchronously invoked operations, and docs should rather state that instead of prohibiting any async calls.